src/mpi/comm/commutil.c

```diff
@@ -35,14 +35,122 @@ MPIR_Object_alloc_t MPIR_Comm_mem = {
 
 /* Communicator creation functions */
 struct MPIR_Commops *MPIR_Comm_fns = NULL;
-struct MPIR_Comm_hint_fn_elt {
-    char name[MPI_MAX_INFO_KEY];
+static int MPIR_Comm_commit_internal(MPIR_Comm * comm);
+
+/* Communicator hint functions */
+/* For balance of simplicity and feature, we'll internally use integers for both keys
+ * and values, and provide facilities to translate from and to string-based infos.
+ */
+
+struct MPIR_HINT {
+    const char *key;
     MPIR_Comm_hint_fn_t fn;
-    void *state;
-    UT_hash_handle hh;
+    int type;
+    int attr;                   /* e.g. whether this key is local */
 };
-static struct MPIR_Comm_hint_fn_elt *MPID_hint_fns = NULL;
-static int MPIR_Comm_commit_internal(MPIR_Comm * comm);
+static struct MPIR_HINT MPIR_comm_hint_list[MPIR_COMM_HINT_MAX];
+static int next_comm_hint_index = MPIR_COMM_HINT_PREDEFINED_COUNT;
+
+int MPIR_Comm_register_hint(int index, const char *hint_key, MPIR_Comm_Comm_hint_fn_t fn,
+                            int type, int attr)
+{
+    if (index == 0) {
+        index = next_comm_hint_index;
+        next_comm_hint_index++;
+        MPIR_Assert(index < MPIR_COMM_HINT_MAX);
+    } else {
+        MPIR_Assert(index > 0 && index < MPIR_COMM_HINT_PREDEFINED_COUNT);
+    }
+    MPIR_comm_hint_list[index] = (struct MPIR_HINT) {
+    hint_key, fn, type, attr};
+    return index;
+}
+
+static int parse_string_value(const char *s, int type, int *val)
+{
+    if (type == MPIR_COMM_HINT_TYPE_BOOL) {
+        if (strcmp(s, "true") == 0) {
+            *val = 1;
+        } else if (strcmp(s, "false") == 0) {
+            *val = 0;
+        } else {
+            *val = atoi(s);
+        }
+    } else if (type == MPIR_COMM_HINT_TYPE_INT) {
+        *val = atoi(s);
+    } else {
+        return -1;
+    }
+    return 0;
+}
+
+static int get_string_value(char *s, int type, int val)
+{
+    if (type == MPIR_COMM_HINT_TYPE_BOOL) {
+        strncpy(s, val ? "true" : "false", MPI_MAX_INFO_VAL);
+    } else if (type == MPIR_COMM_HINT_TYPE_INT) {
+        MPL_snprintf(s, MPI_MAX_INFO_VAL, "%d", val);
+    } else {
+        return -1;
+    }
+    return 0;
+}
+
+int MPII_Comm_set_hints(MPIR_Comm * comm_ptr, MPIR_Info * info)
+{
+    LL_FOREACH(info, curr_info) {
+        if (curr_info->key == NULL)
+            continue;
+        for (int i = 0; i < next_comm_hint_index; i++) {
+            if (MPIR_comm_hint_list[i].key &&
+                strcmp(curr_info->key, MPIR_comm_hint_list[i].key) == 0) {
+                int val;
+                int errno = parse_string_value(curr_info->value, MPIR_comm_hint_list[i].type, &val);
+                if (errno == 0) {
+                    if (MPIR_comm_hint_list[i].fn) {
+                        MPIR_comm_hint_list[i].fn(comm_ptr, i, val);
+                    } else {
+                        comm_ptr->hints[i] = val;
+                    }
+                }
+            }
+        }
+    }
+    /* FIXME: run collective to ensure hints consistency */
```

Comment:
Consistency of non-local (i.e., global) hints right?

Reply:
Yes. Too complicated for right now, so I am pushing it off for a separate PR.